aptos-move/e2e-move-tests/src/tests/function_caches.rs

```diff
@@ -0,0 +1,83 @@
+// Copyright (c) Aptos Foundation
+// Licensed pursuant to the Innovation-Enabling Source Code License, available at https://github.com/aptos-labs/aptos-core/blob/main/LICENSE
+
+//! Tests that interpreter caches frees all allocated data structures for recursively called
+//! functions.
+
+use crate::{assert_success, tests::common, MoveHarness};
+use aptos_framework::BuildOptions;
+use aptos_package_builder::PackageBuilder;
+use aptos_transaction_simulation::Account;
+use aptos_types::transaction::TransactionStatus;
+use move_core_types::account_address::AccountAddress;
+
+#[test]
+fn test_function_caches_for_recursive_functions_do_not_leak_memory() {
+    let mut h = MoveHarness::new();
+    let acc = h.new_account_at(AccountAddress::from_hex_literal("0x99").unwrap());
+
+    let status = publish(
+        &mut h,
+        &acc,
+        r#"
+        module 0x99::m {
+            public entry fun factorial(n: u64) {
+                factorial_impl(n);
+            }
+
+            fun factorial_impl(n: u64): u64 {
+                if (n <= 1) {
+                    1
+                } else {
+                    n * factorial_impl(n - 1)
+                }
+            }
+        }
+        "#,
+    );
+    assert_success!(status);
+
+    let run_factorial = || {
+        for _ in 0..300 {
+            let status = h.run_entry_function(
+                &acc,
+                str::parse("0x99::m::factorial").unwrap(),
+                vec![],
+                vec![bcs::to_bytes(&4_u64).unwrap()],
+            );
+            assert_success!(status);
+        }
+    };
+
+    let measure_memory_growth = || {
+        let start = memory_stats::memory_stats().unwrap().physical_mem;
```

Comment:
Makes sense to measure memory usage, but wonder how reliable this is for testing?